Add initial support for Google's Find My Device network

aaronjamt opened the pull request:
This currently only supports fixed keys, and thus doesn't work with the official Find Device app, although it does work with the unnofficial command-line tool: https://github.com/leonboe1/GoogleFindMyTools

aaronjamt committed:
Initial Google support

aaronjamt committed:
Correct Google data size
The advertisement payload for Google's network should be 29 bytes, not 31.

aaronjamt committed:
Reduce broadcast interval for Google

MatthewKuKanich commented:
Looks good! I'll perform some testing tomorrow and then update documentation on how to use the Google network and it's limitations. Then I'll merge the changes with a few edits for things like clamping broadcast interval and give you credit in the project. Thanks!

MatthewKuKanich reviewed (approved):
Initial testing all works, thanks!

I'll likely do a bit more testing and then merge

gl0wa commented:
@aaronjamt how did you register a Google FMDN tag in FindMyFlipper? I've got a Advertisement Key generated with https://github.com/leonboe1/GoogleFindMyTools but guessing it doesn't just go raw into payload, right?

aaronjamt commented:
> @aaronjamt how did you register a Google FMDN tag in FindMyFlipper? I've got a Advertisement Key generated with https://github.com/leonboe1/GoogleFindMyTools but guessing it doesn't just go raw into payload, right?

There's probably a better way to do this but I believe I used the ESP32 firmware in the repo to make an ESP32 advertise as a Find My device, then used the [nRF Connect](https://play.google.com/store/apps/details?id=no.nordicsemi.android.mcp) app on my phone to grab the actual payload being advertised, then put that into the file on the SD card. I did actually lose that file, though, so I'll see if I can figure out how to do it again and let you know if I can find a better method.

aaronjamt commented:
Looks like the MAC address doesn't matter and the data should be:
```
02 01 06 19 16 AA FE 41 [Advertisement Key] 00
```

This is based on [the ESP32 firmware](https://github.com/leonboe1/GoogleFindMyTools/blob/97bba095e438563e499808bf4018ce76f60bbcd5/ESP32Firmware/main/main.c#L44C1-L55C3):

```
uint8_t adv_raw_data[31] = {
    0x02,   // Length
    0x01,   // Flags data type value
    0x06,   // Flags data
    0x19,   // Length
    0x16,   // Service data data type value
    0xAA,   // 16-bit service UUID
    0xFE,   // 16-bit service UUID
    0x41,   // FMDN frame type with unwanted tracking protection mode indication
            // 20-byte ephemeral identifier (inserted below)
            // Hashed flags (implicitly initialized to 0)
};
```
where the 20-byte ephemeral identifier is the Advertisement Key, and the hashed flags are just set to 0.

gl0wa commented:
This worked! Thanks a lot for your help, and for submitting this PR! ![share_5825152290359749476](https://github.com/user-attachments/assets/f4ad3a2d-23d5-4df6-a254-e86bd3bc1f2c)

aaronjamt commented:
> This worked! Thanks a lot for your help, and for submitting this PR!

I'm glad that worked and happy to help!

---

That photo is very interesting. I didn't think the official Find My app worked, only the Python script. Have you been able to see its location on a map, or any information other than "near Bathroom speaker at Home"?

Also, if you edit [`SpotApi/CreateBleDevice/create_ble_device.py`](https://github.com/leonboe1/GoogleFindMyTools/blob/97bba095e438563e499808bf4018ce76f60bbcd5/SpotApi/CreateBleDevice/create_ble_device.py), you can change the URL for the icon on [line 36](https://github.com/leonboe1/GoogleFindMyTools/blob/97bba095e438563e499808bf4018ce76f60bbcd5/SpotApi/CreateBleDevice/create_ble_device.py#L36) (I used `http://static.tildacdn.com/tild6361-6634-4361-b433-396433333164/fpr-001_light_web_2_.png` for my Flipper) and the manufacturer and model information on [lines 72-73](https://github.com/leonboe1/GoogleFindMyTools/blob/97bba095e438563e499808bf4018ce76f60bbcd5/SpotApi/CreateBleDevice/create_ble_device.py#L72C5-L73C38) (which I set to "Flipper" and "Flipper Zero"). If you want to get extra fancy, I also changed [line 31](https://github.com/leonboe1/GoogleFindMyTools/blob/97bba095e438563e499808bf4018ce76f60bbcd5/SpotApi/CreateBleDevice/create_ble_device.py#L31) to my Flipper's name, so that can also show up, although I believe you can also change that in the Find My app.

gl0wa commented:
>  I didn't think the official Find My app worked 

@aaronjamt ass soon as I generated "Advertisement Key", a new device appeared on my Android  FMDN app. And yes, I had to change it's name (thanks for the tip on how to change the icon!). It was possible to change the name on my Android phone, same, as I would change it for a commercial tag.

But until I set the correct payload on my Flipper, there was no location on that device. Once the payload was right, as soon as I opened Find My Device app, the "home" location was already there.

> Have you been able to see its location on a map

I haven't tested that. Flipper didn't move out of "home" since yet. Will let you know when it does.

Based on few other FMDN tags I have, they always show up this way in the app when at home. Looks like many of Google's speakers and smart displays (we have few of them all around our place) can pickup advertising packets. When this happens they show up like this, but also, sometimes as "near Home" (without specific device).

aaronjamt commented:
> >  I didn't think the official Find My app worked 
> 
> @aaronjamt ass soon as I generated "Advertisement Key", a new device appeared on my Android  FMDN app. And yes, I had to change it's name (thanks for the tip on how to change the icon!). It was possible to change the name on my Android phone, same, as I would change it for a commercial tag.
> 
> But until I set the correct payload on my Flipper, there was no location on that device. 

Right, this is the behavior I'd expect. I don't have any Google Home stuff, just my phone, but when I added my Flipper, it did show up in the app, minus the location until I added the payload to the Flipper.

> Once the payload was right, as soon as I opened Find My Device app, the "home" location was already there.

That's very interesting to me. This does not implement several features required for proper location tracking, in order for it to show up in the app. The Python script used for generating the advertisement payload in your Google account is also used for locating the device, and is the only way I've ever been able to get location information from it. When I check on my phone, it always says "location unavailable" or similar, only the script has been able to fetch my Flipper's location.

> > Have you been able to see its location on a map
> 
> I haven't tested that. Flipper didn't move out of "home" since yet. Will let you know when it does.

I would be very interested to hear what happens here. I'm wondering if the speakers somehow detect the device differently snd are able to show it as "Home" even if the GPS location information isn't sent correctly.

FlixiDoe commented:
Is Google supported now? @aaronjamt @MatthewKuKanich

aaronjamt commented:
> Is Google supported now? @aaronjamt @MatthewKuKanich 

Yes, although I've seen mixed reports about how well it works, so YMMV.

IdontexistVX commented:
> Looks like the MAC address doesn't matter and the data should be:
> 
> ```
> 02 01 06 19 16 AA FE 41 [Advertisement Key] 00
> ```
> 
> This is based on [the ESP32 firmware](https://github.com/leonboe1/GoogleFindMyTools/blob/97bba095e438563e499808bf4018ce76f60bbcd5/ESP32Firmware/main/main.c#L44C1-L55C3):
> 
> ```
> uint8_t adv_raw_data[31] = {
>     0x02,   // Length
>     0x01,   // Flags data type value
>     0x06,   // Flags data
>     0x19,   // Length
>     0x16,   // Service data data type value
>     0xAA,   // 16-bit service UUID
>     0xFE,   // 16-bit service UUID
>     0x41,   // FMDN frame type with unwanted tracking protection mode indication
>             // 20-byte ephemeral identifier (inserted below)
>             // Hashed flags (implicitly initialized to 0)
> };
> ```
> 
> where the 20-byte ephemeral identifier is the Advertisement Key, and the hashed flags are just set to 0.

@aaronjamt 
I am having problems with this, I have an advertisement key, installed the new version with the find hub stuff and can't seem to get it right, here is the file that got generated
[findmy_state.txt](https://github.com/user-attachments/files/26648947/findmy_state.txt)
is there something wrong I am doing, or am I just a dumb ass

aaronjamt commented:
At some point last year it stopped working for me and I never was able to get it working again. I'm not sure if it's no longer usable, or if I'm doing something wrong, but I haven't had a chance to look into it. I'll try to take a look soon.
